client/gui/FramerateManager.cpp

```diff
@@ -11,12 +11,26 @@
 #include "StdInc.h"
 #include "FramerateManager.h"
 
+#include "../../lib/CConfigHandler.h"
+#include <SDL_video.h>
+
 FramerateManager::FramerateManager(int targetFrameRate)
-	: targetFrameTime(Duration(boost::chrono::seconds(1)) / targetFrameRate)
-	, lastFrameIndex(0)
+	: lastFrameIndex(0)
 	, lastFrameTimes({})
-	, lastTimePoint (Clock::now())
+	, lastTimePoint(Clock::now())
 {
+	if(settings["video"]["vsync"].Bool())
+	{
+		static int display_in_use = settings["video"]["displayIndex"].Integer();
+		SDL_DisplayMode mode;
+		SDL_GetCurrentDisplayMode(display_in_use, &mode);
+		int displayRefreshRate = mode.refresh_rate;
+		logGlobal->info("Display refresh rate is %d", displayRefreshRate);
+		targetFrameTime = Duration(boost::chrono::seconds(1)) / displayRefreshRate;
+	} else
+	{
+		targetFrameTime = Duration(boost::chrono::seconds(1)) / targetFrameRate;
+	}
```

Comment:
Why you decided to change the approach?

Previous solution by changing sleep should have been sufficient and more correct.

Reply:
The previous approach led to the animation of the moving hero freezing every second or so.

Reply:
That's a separate bug. Your approach was correct.

Try changing CGuiHandler::renderFrame like this:
```cpp
void CGuiHandler::renderFrame()
{
	{
		boost::recursive_mutex::scoped_lock un(*CPlayerInterface::pim);

		if(nullptr != curInt)
			curInt->update();

		if(settings["video"]["showfps"].Bool())
			drawFPSCounter();

		SDL_UpdateTexture(screenTexture, nullptr, screen->pixels, screen->pitch);

		SDL_RenderClear(mainRenderer);
		SDL_RenderCopy(mainRenderer, screenTexture, nullptr, nullptr);

		CCS->curh->render();
		windows().onFrameRendered();
	}
	SDL_RenderPresent(mainRenderer);
	framerate().framerateDelay(); // holds a constant FPS
}

Problem is - game maintains lock on "CPlayerInterface::pim" while wating for vsync, blocking any other actions such as client<->server communication